service/history/archival_queue_task_executor.go

```diff
@@ -0,0 +1,349 @@
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/service/history/workflow"
+	"go.temporal.io/server/service/history/workflow/cache"
+)
+
+// NewArchivalQueueTaskExecutor creates a new queue task executor for the archival queue.
+// If you use this executor, you must monitor for any metrics.ArchivalTaskInvalidURI errors.
+// If this metric is emitted, it means that an archival URI is invalid and the task will never succeed, which is a
+// serious problem because the archival queue retries tasks forever.
+func NewArchivalQueueTaskExecutor(
+	archiver archival.Archiver,
+	shardContext shard.Context,
+	workflowCache cache.Cache,
+	relocatableAttributesFetcher workflow.RelocatableAttributesFetcher,
+	metricsHandler metrics.MetricsHandler,
+	logger log.Logger,
+) queues.Executor {
+	return &archivalQueueTaskExecutor{
+		archiver:                     archiver,
+		shardContext:                 shardContext,
+		workflowCache:                workflowCache,
+		relocatableAttributesFetcher: relocatableAttributesFetcher,
+		metricsHandler:               metricsHandler,
+		logger:                       logger,
+	}
+}
+
+// archivalQueueTaskExecutor is an implementation of queues.Executor for the archival queue.
+type archivalQueueTaskExecutor struct {
+	archiver                     archival.Archiver
+	shardContext                 shard.Context
+	workflowCache                cache.Cache
+	metricsHandler               metrics.MetricsHandler
+	logger                       log.Logger
+	relocatableAttributesFetcher workflow.RelocatableAttributesFetcher
+}
+
+// Execute executes a task from the archival queue.
+func (e *archivalQueueTaskExecutor) Execute(
+	ctx context.Context,
+	executable queues.Executable,
+) (tags []metrics.Tag, isActive bool, err error) {
+	task := executable.GetTask()
+	taskType := queues.GetArchivalTaskTypeTagValue(task)
+	tags = []metrics.Tag{
+		getNamespaceTagByID(e.shardContext.GetNamespaceRegistry(), task.GetNamespaceID()),
+		metrics.TaskTypeTag(taskType),
+	}
+	switch task := task.(type) {
+	case *tasks.ArchiveExecutionTask:
+		err = e.processArchiveExecutionTask(ctx, task)
+	default:
+		err = fmt.Errorf("task with invalid type sent to archival queue: %+v", task)
+	}
+	return tags, true, err
+}
+
+// processArchiveExecutionTask processes a tasks.ArchiveExecutionTask
+// First, we load the mutable state to populate an archival.Request.
+// Second, we unlock the mutable state and send the archival request to the archival.Archiver.
+// Finally, we lock the mutable state again, and send a deletion follow-up task to the history engine.
+func (e *archivalQueueTaskExecutor) processArchiveExecutionTask(ctx context.Context, task *tasks.ArchiveExecutionTask) (err error) {
+	logger := log.With(e.logger, tag.Task(task))
+	request, err := e.getArchiveTaskRequest(ctx, logger, task)
+	if err != nil {
+		if err == ErrMutableStateIsNil || err == ErrWorkflowExecutionIsStillRunning {
```

Comment:
nit: either handle at the top level in Execute(...) or from the return function loadAndVersionCheckMutableState